### ISSUE TITLE
Use alpine as the base image (again)

### ISSUE THREAD

**Issue by 1138-4EB** [wagoodman/dive#243]
A month ago (https://github.com/wagoodman/dive/commit/c83bc9e3d03356a942a90b9a6c9832e336363229#diff-3a895417c300af3c34fab6ebb556e984), `CGO_ENABLED=0` was removed from `.goreleaser.yml` because some dependencies of podman do require dynamic loading. I.e., it is no longer possible to build `dive` statically. As a result, binary files built on Ubuntu of Fedora host won't work in Alpine containers. So, Alpine based images were dropped in favour of `debian:sid-slim` (https://github.com/wagoodman/dive/commit/65f87b00a947a20dc62f57cfef1c59009ba2fec2#r35804708). However, these images are 10-15 times larger than previous `slim` variants.

A possible workaround (suggested in https://github.com/wagoodman/dive/commit/65f87b00a947a20dc62f57cfef1c59009ba2fec2#r35815381) is to convert podman support into a plugin, so that static compilation is possible for users that do not need it. However, this would be cumbersome for maintenance.

Instead, I suggest executing `goreleaser` twice. Once on an Ubuntu or Fedora host in order to generate `deb` and `rpm` artifacts. Then, inside a `golang:alpine` container, in order to generate docker images. Feature-wise, all artifacts would be equivalent; the only difference being that some depend on `glib` and others on `musl`.

Build dependencies in a `golang:alpine` container are `gpgme-dev gcc musl-dev btrfs-progs-dev lvm2-dev` and runtime dependencies are `gpgme device-mapper`. For example:

```Dockerfile
ARG ALPINE_VERSION="3.10"
ARG DOCKER_CLI_VERSION="19.03.1"

FROM alpine:$ALPINE_VERSION AS docker
ARG DOCKER_CLI_VERSION
RUN apk -U --no-cache add curl \
 && curl -L https://download.docker.com/linux/static/stable/x86_64/docker-$DOCKER_CLI_VERSION.tgz | tar -xzf - docker/docker --strip-component=1 -C /tmp

FROM alpine:$ALPINE_VERSION
COPY --from=docker /tmp/docker /usr/local/bin
COPY dive /dive
RUN apk add -U --no-cache gpgme device-mapper
ENTRYPOINT ["/dive"]
```

```sh
#!/usr/bin/env sh

set -e

cd "$(dirname $0)"

docker build -t wagoodman/dive:golang- <<EOF
FROM golang:alpine AS build
RUN apk add -U --no-cache gpgme-dev gcc musl-dev btrfs-progs-dev lvm2-dev
EOF

docker run --rm -v /$(pwd)://src -w //src wagoodman/dive:golang sh -c "go build -o slim/dive -ldflags '-s -w'"

docker build -t wagoodman/dive:slim .
```

**Comment by wagoodman** [comment]
@1138-4EB Thanks for your help this week! The 0.9.0 was rough to say the least, but I recently released 0.9.1 that should accommodate both the dependency problems and the base image. Given the state of 0.9.0, building twice as you suggested would be a great hold-over approach while I found a longer term solution (such as the plugin approach). 

Though, your comment about possibly migrating to a plugin system gave me an I idea: Podman is already a separate binary with `image save` CLI options. Though I didn't go this route for Docker, for Podman, I can shell out to a process and stream the results back. This is the approach that 0.9.1 is taking (https://github.com/wagoodman/dive/commit/3fbc016826b4a1582acbb91c103ca094216a6cd7). I'm not in love with the approach, but I really like the end result: keeping libpod dependencies out of dive. This easily allowed moving back to alpine (https://github.com/wagoodman/dive/commit/bfcfc54ee34b623274a91ca48379ac229b5415c2)... I'll close this issue since we've pivoted back to alpine (and will hopefully stay there!)

**Comment by 1138-4EB** [comment]
> @1138-4EB Thanks for your help this week!

You are welcome! This is a really cool project, and I'm glad to help within my limitations.

> Given the state of 0.9.0, building twice as you suggested would be a great hold-over approach while I found a longer term solution (such as the plugin approach).

We will see how it goes. But I think that you should consider keeping two build configs/workflows even after you find a longer term solution. This is because, unlike native binaries, which can require being tailored to the (multiple) targets, you can always use a single docker image on any target. I.e., for native binaries, you might want to add metadata and be more careful with packaging, since those are expected to be installed on hosts which you don't know. However, since you can decide every bit of the image, there will always be room for simplification/optimization.

More precisely, I'm considering to add multiarch docker images, so that dive is easily usable on devices such as Raspberry Pi, Rock960, or on PPC64 servers. While this is relatively straightforward using docker images/containers, I don't think it is worth doing it for native binaries, due to the size of the resulting target matrix. Should you/we want to support native binaries for multiple architectures, I think we'd better use services such as COPR.

> I'm not in love with the approach, but I really like the end result: keeping libpod dependencies out of dive. This easily allowed moving back to alpine ([bfcfc54](https://github.com/wagoodman/dive/commit/bfcfc54ee34b623274a91ca48379ac229b5415c2))... I'll close this issue since we've pivoted back to alpine (and will hopefully stay there!)

As commented in https://github.com/wagoodman/dive/issues/242#issuecomment-552126588, I'm really glad that you reconsidered the solution. I understand that you are not in love with the current approach, but I think that your fast reaction deserves a sincere acknowledgement: thanks! There are many other larger and very used projects which are broken for months; hence, it is really grateful to find so responsive maintainers.

Unfortunately, I'm not experienced with actually implementing plugin systems for golang. You might want to open an issue about it, so that other users can help providing hints about it.

> Podman is already a separate binary with `image save` CLI options. Though I didn't go this route for Docker, for Podman, I can shell out to a process and stream the results back.

Just because of curiosity why didn't you take this approach for Docker? Regarding the conversation in https://github.com/wagoodman/dive/commit/65f87b00a947a20dc62f57cfef1c59009ba2fec2#r34785830, which lead to opening #226, it seems that you are using the binary. Note that I have not diven into dive's codebase yet, so I'm still blind.

I'm asking this because I think there are two types of plugin approaches:

- Build time plugins. Developers can select which features to enable when building a binary. Dependencies are fixed and multiple binaries need to be generated for different plugin sets.
- Runtime plugins. Developers build a single `dive` binary and multiple plugin libraries (`.a`, `.so`, `.dll`, whatever). Users that want basic features only will install `dive`. Users which want other features, will add the plugin files to a preset location.

In the second case, it might be discussed how slim should `dive` be. I.e., should docker support be a plugin too?
